### PR TITLE
Move dispersion fitter implementation to components

### PR DESCRIPTION
**Pull request by caseyflex**
This PR is exploring an intermediate approach between two old PRs, neither of which was merged:
https://github.com/flexcompute/tidy3d/pull/1773
https://github.com/flexcompute/tidy3d/pull/1809

The first PR moves the fitter to components. The second PR separates it out to a new package.

In this PR, the core fitting algorithm is moved to components, but the fitter plugin with the quality-of-life features stays in plugins.

### CODE REVIEW

File: tidy3d/components/dispersion_fitter.py [hunk 1]
```diff
@@ -0,0 +1,936 @@
...
+OMEGA_POLE_CLOSE_ATOL = 1e-10
+
+# Range for checking upper bound of Im[eps], in addition to extrema method.
+# The range is in unit of eV and it's in log scale.
+LOSS_CHECK_MIN = -10
+LOSS_CHECK_MAX = 4
+LOSS_CHECK_NUM = 1000
+
+
+def imag_ep_extrema(poles: ArrayComplex1D, residues: ArrayComplex1D) -> ArrayFloat1D:
```
Comment (tylerflex) [inline tidy3d/components/dispersion_fitter.py]: what is `ep`? might want to consider giving this function a more intuitive name.

**Review by tylerflex — COMMENTED** [review]
Looks good to me. I think having it like this (inside of components, but somewhat separated) makes sense for the time being.

**Review by tylerflex — APPROVED** [review]

File: tidy3d/plugins/dispersion/fit_fast.py [hunk 2]
```diff
@@ -698,147 +91,26 @@ def fit(
...
-                "Dispersion fitter cannot have 'max_num_poles' less than 'min_num_poles'."
-            )
-
-        omega = PoleResidue.angular_freq_to_eV(PoleResidue.Hz_to_angular_freq(self.freqs[::-1]))
-        eps = self.eps_data[::-1]
-
-        init_model = FastFitterData.initialize(
-            omega, eps, eps_inf, advanced_param or AdvancedFastFitterParam()
+        omega_data = PoleResidue.Hz_to_angular_freq(self.freqs[::-1])
+        eps_data = self.eps_data[::-1]
```
Comment (weiliangjin2021) [inline tidy3d/plugins/dispersion/fit_fast.py]: Unrelated question to this PR: does fitter only work with descending frequency list?

**Review by weiliangjin2021 — APPROVED** [review]
Looks nice!

File: tidy3d/plugins/dispersion/fit_fast.py (hunk 2 above)
Comment (caseyflex) [reply tidy3d/plugins/dispersion/fit_fast.py]: Good call, it should work regardless of the order. Let me remove this `[::-1]`. I think this was an artifact from early testing and debugging of the fitter.

**Comment by caseyflex** [comment]
@tylerflex revised. can we merge this?

**Comment by tylerflex** [comment]
yes, but maybe this is best for 2.8?  @momchil-flex

**Comment by momchil-flex** [comment]
Yeah let's do 2.8 please.

**Comment by momchil-flex** [comment]
Also note that it shouldn't be moving the notebooks submodule state unless you explicitly intend to.

**Pull request merged**